src/cwe_checker_lib/src/analysis/taint/mod.rs

```diff
@@ -107,19 +118,98 @@ pub trait TaintAnalysis<'a>: HasCfg<'a> + HasVsaResult<PiData> + AsRef<Project>
-    fn update_return(
+    /// Returns an empty state, i.e., information is propagated through the call
+    /// but the analysis stays intraprocedural.
+    fn update_return_callee(
         &self,
         _state: &State,
         _call_term: &Term<Jmp>,
         _return_term: &Term<Jmp>,
         _calling_convention: &Option<String>,
-    ) {
+    ) -> Option<State> {
+        Some(State::new_empty())
+    }
+
+    /// Corresponds to returns from calls to other functions within the program.
+    ///
+    /// By implementing this method you can perform interprocedural taint
+    /// analysis. See
+    /// [`forward_interprocedural_fixpoint::Context::update_return`] for more
+    /// information.
+    ///
+    /// # Default
+    ///
+    /// Depending on the availability of `state_before_call` and
+    /// `state_before_return` the return value is computed according to the
+    /// following scheme:
+    ///
+    /// ```table
+    /// | state_before_call/    |                                                       |                     |
+    /// |   state_before_return | Some                                                  | None                |
+    /// |-----------------------|-------------------------------------------------------|---------------------|
+    /// | Some                  | Some(update_call_generic.merge(update_return_callee)) | update_call_generic |
+    /// |                       | IF both are Some ELSE None                            |                     |
+    /// |-----------------------|-------------------------------------------------------|---------------------|
+    /// | None                  | update_return_callee                                  | None                |
+    /// ```
+    fn update_return(
+        &self,
+        state_before_return: Option<&State>,
+        state_before_call: Option<&State>,
+        call_term: &Term<Jmp>,
+        return_term: &Term<Jmp>,
+        calling_convention: &Option<String>,
+    ) -> Option<State> {
+        match (state_before_call, state_before_return) {
+            (Some(state_before_call), Some(state_before_return)) => {
+                let state_from_caller =
+                    self.update_call_generic(state_before_call, &call_term.tid, calling_convention);
+                let state_from_callee = self.update_return_callee(
+                    state_before_return,
+                    call_term,
+                    return_term,
+                    calling_convention,
+                );
+
+                match (state_from_caller, state_from_callee) {
+                    (Some(state_caller), Some(state_callee)) => {
+                        Some(state_caller.merge(&state_callee))
+                    }
+                    // If one implementation indicated that no information
+                    // should be propagated by returning `None` we ignore what
+                    // the other call returned.
+                    _ => None,
+                }
```

Comment:
Well, in #451 you basically implemented my suggestion (in `update_return_callee` instead of here). But my issue is that naive usage of `update_return_callee`, i.e. just propagate the unmodified taint state, will lead to soundness errors. So we can make the behavior of #451 the standard behavior (in `update_return`, so that just overwriting `update_return_callee` is always safe) or we need to put big warning signs into the doc comments of `update_return` and `update_return_callee` to make implementers aware of the possible foot gun.

... Maybe discussing the issue in the doc comments is something we should do anyway.

Reply:
Yes, I would unconditionally remove memory taint from the return value of `update_return_callee` in `update_return` to make sure that naive implementations of `update_return_callee` do not lead to unsoundness. Panicking would also be an alternative, although it would make the standard implementation slightly less useful. On the other hand, it would force implementers to think about the issue, which might be a good thing?

Regarding code reviews: In my experience, PRs for the cwe_checker often tend to be quite large. So reviewers are happy about everything that they do not have to remember during the review. ;-) The big doc comments are a good idea anyway.